p2p/cuprate-p2p/src/block_downloader.rs

```diff
@@ -0,0 +1,1181 @@
+
+    /// The running block download tasks.
+    ///
+    /// Returns:
+    /// - The start height of the batch
+    /// - A result contains the batch or an error.
+    #[allow(clippy::type_complexity)]
+    block_download_tasks: JoinSet<(
+        u64,
+        Result<(ClientPoolDropGuard<N>, BlockBatch), BlockDownloadError>,
+    )>,
+    /// The running chain entry tasks.
+    ///
+    /// Returns a result of the chain entry or an error.
+    #[allow(clippy::type_complexity)]
+    chain_entry_task: JoinSet<Result<(ClientPoolDropGuard<N>, ChainEntry<N>), BlockDownloadError>>,
+
+    /// The current inflight requests.
+    ///
+    /// This a map of batch start heights to block ids and related information of the batch.
+    inflight_requests: BTreeMap<u64, BlocksToRetrieve<N>>,
+
+    /// A queue of ready batches.
+    ready_batches: BinaryHeap<ReadyQueueBatch>,
+    /// The size, in bytes, of all the batches in `ready_batches`.
+    ready_batches_size: usize,
+
+    /// A queue of failed batches start height's that should be retried.
+    ///
+    /// Wrapped in [`Reverse`] so we prioritize early batches.
+    failed_batches: BinaryHeap<Reverse<u64>>,
+
+    /// The [`BufferAppender`] that gives blocks to Cuprate.
+    buffer_appender: BufferAppender<BlockBatch>,
+
+    /// The [`BlockDownloaderConfig`].
+    config: BlockDownloaderConfig,
+}
+
+impl<N: NetworkZone, S, C> BlockDownloader<N, S, C>
+where
+    S: PeerSyncSvc<N> + Clone,
+    C: Service<ChainSvcRequest, Response = ChainSvcResponse, Error = tower::BoxError>
+        + Send
+        + 'static,
+    C::Future: Send + 'static,
+{
+    /// Creates a new [`BlockDownloader`]
+    fn new(
+        client_pool: Arc<ClientPool<N>>,
+
+        peer_sync_svc: S,
+        our_chain_svc: C,
+        buffer_appender: BufferAppender<BlockBatch>,
+
+        config: BlockDownloaderConfig,
+    ) -> Self {
+        BlockDownloader {
+            client_pool,
+            pending_peers: BTreeMap::new(),
+            peer_sync_svc,
+            our_chain_svc,
+            amount_of_blocks_to_request: config.initial_batch_size,
+            amount_of_blocks_to_request_updated_at: 0,
+            amount_of_empty_chain_entries: 0,
+            block_download_tasks: JoinSet::new(),
+            chain_entry_task: JoinSet::new(),
+            inflight_requests: BTreeMap::new(),
+            ready_batches: BinaryHeap::new(),
+            ready_batches_size: 0,
+            failed_batches: BinaryHeap::new(),
+            buffer_appender,
+            config,
+        }
+    }
+
+    /// Checks if we can make use of any peers that are currently pending requests.
+    async fn check_pending_peers(&mut self, chain_tracker: &mut ChainTracker<N>) {
+        tracing::debug!("Checking if we can give any work to pending peers.");
+
+        // HACK: The borrow checker doesn't like the following code if we don't do this.
+        let mut pending_peers = mem::take(&mut self.pending_peers);
```

Comment:
Usually this indicates some `struct` or `fn` needs to be factored but maybe not possible here? Or at least unwieldy.

At least `BTreeMap::default()` doesn't allocate.
```suggestion
        // FIXME: We need partial mutable access to `self` (not possible) so
        // swap with an empty `BTreeMap` for the remainder of the function.
        // <https://github.com/rust-lang/rfcs/issues/1215>
        let mut pending_peers = mem::take(&mut self.pending_peers);
```

Reply:
Yeah thats what I had to do with the chain tracker, I thought it would be annoying to do it to `pending_peers` as well but it isn't actually too bad